src/Tooltip/Tooltip.js

```diff
@@ -4,91 +4,173 @@ import ReactDOM from 'react-dom';
 
 import BindMixin from '../Mixin/BindMixin';
 import DOMUtil from '../Util/DOMUtil';
+import Portal from '../Portal/Portal';
 import Util from '../Util/Util';
 
+const ARROW_SIZE = 7;
+
 class Tooltip extends Util.mixin(BindMixin) {
   get methodsToBind() {
     return [
+      'dismissTooltip',
       'getIdealLocation',
       'handleMouseEnter',
-      'handleMouseLeave'
+      'handleMouseLeave',
+      'handleTooltipMouseEnter',
+      'handleTooltipMouseLeave'
     ];
   }
 
   constructor() {
-    super();
-
+    super(...arguments);
     this.state = {isOpen: false};
   }
 
-  getAnchor(anchor, position, renderedPosition, viewportWidth, viewportHeight) {
-    // Change the anchor if the tooltip will be rendered off the screen.
-    if (position === 'right' || position === 'left') {
-      if (renderedPosition.top < 0) {
-        anchor = 'start';
-      } else if (renderedPosition.bottom > viewportHeight) {
-        anchor = 'end';
-      }
-    } else if (position === 'top' || position === 'bottom') {
-      if (renderedPosition.right > viewportWidth) {
-        anchor = 'end';
-      } else if (renderedPosition.left < 0) {
-        anchor = 'start';
-      }
+  componentWillUnmount() {
+    this.removeScrollListener();
+  }
+
+  handleMouseEnter(currentAnchor, currentPosition) {
+    let {anchor, position, coordinates} = this.getIdealLocation(currentAnchor,
+      currentPosition);
+    this.setState({anchor, isOpen: true, position, coordinates});
+    this.addScrollListener();
+  }
+
+  handleMouseLeave() {
+    this.dismissTooltip();
+  }
+
+  handleTooltipMouseEnter() {
+    if (this.props.interactive) {
+      this.setState({isOpen: true});
```

Comment:
do we need to readd the listener here? or does `handleMouseLeave` never get called at this point?